### DISCUSSION TITLE
InsertMany then Find will raise duplicate key value violates unique constraint

### DISCUSSION THREAD

**Discussion by dungkaka** [mikro-orm/mikro-orm#4692] [General]
**Describe the bug**
I use insertMany to insert multiple record, but if I find right after insertMany, it cause duplicate key value violates unique constraint.

Code describe
```Typescript
await this.em.insertMany(
    City,
    mongoCities.map((city) => this.em.create(City, city as any))
);

// Stop here will not raise error
// But if continute run below find code, it will raise duplicate key value violates unique constraint

const cities = await this.em.find(City, {});

```

**Expected behavior**
After insertMany, I can find without cause error !

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| mikro-orm | 5.8 |

**Comment by dungkaka** [comment]
Give many thanks to this author of awesome lib.
Btw, I also want to know the difference between these two ways to insertMany, about what happens inside, performance? :

1. Insert many created entity.

```Typescript
await this.em.insertMany(
          City,
          mongoCities.map((city) => this.em.create(City, city as any))
      );

```

2. Persist multiple created entity then flush

```Typescript
mongoCities.forEach((city) => {
    const _city = this.em.create(City, city as any);
    this.em.persist(_city);
});
await this.em.flush();

```

**Comment by B4nan** [comment]
This is not reproducible, sounds like you are using different EM forks (= different contexts), and [auto-flush](https://mikro-orm.io/docs/unit-of-work#flush-modes) gets triggered by the `em.find()`. WIll be happy to help if you provide complete repro, but this is a misusage most probably.

Note that you are not using the method correctly, this is how it's supposed to be used:

```ts
// no need to create entity instances yourseft
await this.em.insertMany(City, mongoCities);

// or create the instances and provide them in the first parameter
const entities = mongoCities.map((city) => this.em.create(City, city as any));
await this.em.insertMany(entities);
```

Btw, you shouldn't need the `as any` cast, sounds like either wrong input type or wrong entity definition.

The difference between `insertMany` and flushing is that `insertMany` won't affect the context anyhow, it's a different approach. On the other hand, using `em.create()` will automatically call `persist` on the just created entity, and as such, flushing will insert that too - that is probably what is happening to you, as you use `em.create()`, leave the entity in the context, and instead of flushing, you use `insertMany` - that just issues the insert query, but keeps the context populated, and the following find call triggers a flush for those. I guess we could ensure the right state on the entity instances after you do the `insertMany`, but it is still a misusage imo.

---

Actually, yes, this is exactly what is happening to you, you are just mixing two approaches. This should help: https://github.com/mikro-orm/mikro-orm/commit/94eed5e1aa7192d1b0658eb5b0cc3589cfaacff5?w=1
